wip: add contentful cms guide (#1467)
withastro/docs: src/pages/en/guides/cms/contentful.md

---
title: Contentful & Astro
description: How to connect your Contentful CMS to your Astro site.

> layout: ~/layouts/CmsGuideLayout.astro

Reviewer: ```suggestion
layout: ~/layouts/MainLayout.astro
```
I'm just going to use the main layout for now so that we can see a Netlify preview of this as we work, too.

---

[Contentful](https://www.contentful.com/) is a headless CMS that allows you to manage content, integrate with other services, and publish to multiple platforms. It is a great choice for Astro projects that need to pull in content from a CMS.

## Prerequisites

> - **An Astro project** - If you don't have an Astro project yet, our [Installation guide](/install/auto/) will get you up and running in no time.

Reviewer: ```suggestion
- **An Astro project** - If you don't have an Astro project yet, our [Installation guide](/en/install/auto/) will get you up and running in no time.
```
link

> - **An Astro project** - If you don't have an Astro project yet, our [Installation guide](/en/install/auto/) will get you up and running in no time.

Reviewer 2: Maybe something like "for this walkthrough, you can choose `Empty Project` as the starter template."

- **Contentful credentials** - Find the following credentials in your Contentful space dashboard **Settings > API keys**:

  - **Contentful space ID** - The ID of your Contentful space. 
  - **Contentful delivery access token** - The access token to consume _published_ content from your Contentful space.
  - **Contentful preview access token** - The access token to consume _unpublished_ content from your Contentful space.

## Integrating Contentful with Astro

### Setting up credentials

To add your contentful credentials to Astro, create an `.env` file in the root of your project and add the following:

```ini title=".env"
CONTENTFUL_SPACE_ID=YOUR_SPACE_ID
CONTENTFUL_DELIVERY_TOKEN=YOUR_DELIVERY_TOKEN
CONTENTFUL_PREVIEW_TOKEN=YOUR_PREVIEW_TOKEN
```

if you would like to have IntelliSense for your Contentful environment variables, you can create a `env.d.ts` file in the `src/` directory and configure `ImportMetaEnv` like this:

```ts title="src/env.d.ts"
interface ImportMetaEnv {
  readonly CONTENTFUL_SPACE_ID: string;
  readonly CONTENTFUL_DELIVERY_TOKEN: string;
  readonly CONTENTFUL_PREVIEW_TOKEN: string;
}
```

> ### Installing the Contentful

Reviewer 2: ```suggestion
### Installing the Contentful Client
```

Install the `contentful` package:

> ```bash

Reviewer 2: We can use the `PackageManagerTabs` component here

npm install contentful
```

> Next, create a new file called `contentful.ts` in your project's `src/lib` directory. This file will create a contenful client that will query your Contentful space using your credentials.

Reviewer 2: ```suggestion
Next, create a new file called `contentful.ts` in a `src/lib` directory in your project.
```
(to clarify that it's a new directory)

  space: import.meta.env.CONTENTFUL_SPACE_ID,
  accessToken: import.meta.env.DEV
    ? import.meta.env.CONTENTFUL_PREVIEW_TOKEN
    : import.meta.env.CONTENTFUL_DELIVERY_TOKEN,
  host: import.meta.env.DEV ? "preview.contentful.com" : "cdn.contentful.com",
});

export default client;
```

The above code snippet is creating a new contentful client and passing in the credentials from the `.env` file. In development mode (`import.meta.env.DEV = true`), the client will use the preview token and preview host. In production mode, the client will use the delivery token and delivery host.

:::caution
While in development mode, your content will be fetched from the **Contentful preview API**. This means that you will be able to see unplublished content from the Contentful web app. At built time, your content will be fetched from the **Contentful delivery API**. This means that only published content will be available in your Astro site. 
:::

## Fetching data from Astro components

Now that we have our Contentful client set up, you can fetch data from Contentful inside your Astro components. 

```astro
---
import contentfulClient from "../lib/contentful";

> // more fetching options in https://contentful.github.io/contentful.js/contentful/9.1.34/ContentfulClientAPI.html

Reviewer 2: Since you link to this in the guide itself (where you get a clickable link), I think we can remove it from the code sample.

> const { items } = await contentfulClient.getEntries();

Reviewer 2: ```suggestion
const { items } = await contentfulClient.getEntries<{ title: string }>();
```
This fixes the type error!
